File: 303. Range Sum Query - Immutable/303.py
class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        self.acc = []
        for i in range(len(nums)):
            if i == 0:
                self.acc.append(nums[i])
            else:
                self.acc.append(nums[i] + self.acc[i-1])

    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        if i == 0:
            return self.acc[j]
        else:
            return self.acc[j] - self.acc[i-1]

if __name__ == "__main__":
    s = NumArray([-2,0,3,-5,2,-1])
    print(s.sumRange(0,2))
    print(s.sumRange(2,5))
    print(s.sumRange(0,5))

# Your NumArray object will be instantiated and called as such:
# numArray = NumArray(nums)
# numArray.sumRange(0, 1)
# numArray.sumRange(1, 2)

File: 303. Range Sum Query - Immutable/303.cpp
class NumArray {
public:
    NumArray(vector<int> nums) {
        sum_ = vector<int>(nums.size(), 0);
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
            sum_[i] = sum;
        }
    }
    
    // [i, j]
    int sumRange(int i, int j) {
        if (i == 0) {
            return sum_[j];
        } else {
            return sum_[j] - sum_[i - 1];
        }
    }
private:
    vector<int> sum_;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */
